File: summary.py
import os
import time
import json
import logging
from datetime import datetime, timedelta
from typing import List, Dict
import pytz
from google.oauth2 import service_account
from googleapiclient.discovery import build
from notion_client import Client
import threading
from queue import Queue
from youtube_transcript_api import YouTubeTranscriptApi

# 定数
MAX_DESCRIPTION_LENGTH = 1000
YOUTUBE_SCOPES = [
    "https://www.googleapis.com/auth/youtube.readonly",
    "https://www.googleapis.com/auth/youtube.force-ssl"
]

def lambda_handler(event, context):
    try:
        main()
        return {
            'statusCode': 200,
            'body': json.dumps('Success')
        }
    except Exception as e:
        return {
            'statusCode': 500,
            'body': json.dumps(str(e))
        }

class VideoInfo:
    def __init__(self, video_id: str, title: str, description: str, published_at: datetime, 
                 channel_title: str, url: str, captions: List[Dict] = None):
        self.video_id = video_id
        self.title = title
        self.description = description
        self.published_at = published_at
        self.channel_title = channel_title
        self.url = url
        self.captions = captions or []

class CaptionInfo:
    def __init__(self, language: str, text: str, is_automatic: bool):
        self.language = language
        self.text = text
        self.is_automatic = is_automatic

def truncate_description(description: str) -> str:
    """説明文を制限する関数"""
    if len(description) > MAX_DESCRIPTION_LENGTH:
        return description[:MAX_DESCRIPTION_LENGTH-3] + "..."
    return description

def get_service_account_client():
    """サービスアカウント認証クライアントの取得"""
    try:
        credentials = service_account.Credentials.from_service_account_file(
            'service-account.json',
            scopes=YOUTUBE_SCOPES
        )
        return build('youtube', 'v3', credentials=credentials)
    except Exception as e:
        logging.error(f"サービスアカウントクライアントの作成に失敗: {e}")
        raise

def get_captions(video_id: str) -> List[CaptionInfo]:
    """字幕情報の取得"""
    try:
        # 利用可能な字幕リストを取得
        transcript_list = YouTubeTranscriptApi.list_transcripts(video_id)
        captions = []

        # 日本語字幕を優先的に取得
        try:
            transcript = transcript_list.find_transcript(['ja'])
            transcript_data = transcript.fetch()
            
            # 字幕テキストを結合
            caption_text = ""
            for item in transcript_data:
                text = item['text'].replace('\n', ' ')
                caption_text += text + " "

            caption_info = CaptionInfo(
                language="ja",
                text=caption_text,
                is_automatic=transcript.is_generated
            )
            captions.append(caption_info)
            
        except:
            # 日本語字幕が無い場合は英語字幕を取得
            try:
                transcript = transcript_list.find_transcript(['en'])
                transcript_data = transcript.fetch()
                
                caption_text = ""
                for item in transcript_data:
                    text = item['text'].replace('\n', ' ')
                    caption_text += text + " "

                caption_info = CaptionInfo(
                    language="en",
                    text=caption_text,
                    is_automatic=transcript.is_generated
                )
                captions.append(caption_info)
                
            except Exception as e:
                logging.warning(f"英語字幕の取得に失敗: {e}")

        return captions

    except Exception as e:
        logging.warning(f"字幕取得エラー (video_id: {video_id}): {e}")
        return []


def get_latest_videos(youtube_service, channel_id: str) -> List[VideoInfo]:
    """最新の動画を取得"""
    try:
        channel_response = youtube_service.channels().list(
            part="contentDetails",
            id=channel_id
        ).execute()

        if not channel_response['items']:
            raise ValueError("チャンネルが見つかりません")

        uploads_playlist_id = channel_response['items'][0]['contentDetails']['relatedPlaylists']['uploads']
        
        playlist_response = youtube_service.playlistItems().list(
            part="snippet",
            playlistId=uploads_playlist_id,
            maxResults=50
        ).execute()

        videos = []
        jst = pytz.timezone('Asia/Tokyo')
        now = datetime.now(jst)
        today = now.replace(hour=0, minute=0, second=0, microsecond=0)
        yesterday = today - timedelta(days=1)

        for item in playlist_response['items']:
            snippet = item['snippet']
            published_at = datetime.fromisoformat(snippet['publishedAt'].replace('Z', '+00:00'))
            published_at_jst = published_at.astimezone(jst)
            published_date = published_at_jst.replace(hour=0, minute=0, second=0, microsecond=0)

            if published_date in (today, yesterday):
                video = VideoInfo(
                    video_id=snippet['resourceId']['videoId'],
                    title=snippet['title'],
                    description=truncate_description(snippet['description']),
                    published_at=published_at,
                    channel_title=snippet['channelTitle'],
                    url=f"https://www.youtube.com/watch?v={snippet['resourceId']['videoId']}"
                )
                videos.append(video)

        return videos

    except Exception as e:
        logging.error(f"動画取得エラー: {e}")
        raise

def split_text(text: str, max_length: int = 2000) -> List[str]:
    """
    テキストを指定された最大長で分割する
    - 文章の途中で分割されないように、文末で区切る
    - 改行や句点で区切られた段落を考慮する
    - 長すぎる文章は強制的に分割する
    """
    if not text:
        return []
    
    result = []
    current_chunk = ""
    
    # 改行で段落に分割
    paragraphs = text.split('\n')
    
    for paragraph in paragraphs:
        if not paragraph.strip():
            continue
            
        # 文章を句点で分割
        sentences = paragraph.split('。')
        
        for sentence in sentences:
            if not sentence.strip():
                continue
                
            # 文が非常に長い場合は強制的に分割
            if len(sentence) > max_length:
                # 現在のチャンクがある場合は追加
                if current_chunk:
                    result.append(current_chunk)
                    current_chunk = ""
                
                # 長い文を強制的に分割
                for i in range(0, len(sentence), max_length):
                    chunk = sentence[i:i + max_length]
                    result.append(chunk)
                continue
            
            # 通常の文処理
            sentence_with_period = sentence + ('。' if sentence != sentences[-1] else '')
            
            # 現在のチャンクに追加可能か確認
            if len(current_chunk + sentence_with_period) <= max_length:
                current_chunk += sentence_with_period
            else:
                # 現在のチャンクを結果に追加
                if current_chunk:
                    result.append(current_chunk)
                current_chunk = sentence_with_period
    
    # 最後のチャンクを追加
    if current_chunk:
        result.append(current_chunk)
    
    return result

def save_to_notion(notion_client, database_id: str, video: VideoInfo):
    """Notionにデータを保存"""
    try:
        # 重複チェック
        results = notion_client.databases.query(
            database_id=database_id,
            filter={
                "property": "URL",
                "rich_text": {
                    "contains": video.video_id
                }
            }
        )

        if results['results']:
            logging.info(f"スキップ: 動画 {video.video_id} は既にNotionに存在します")
            return

        # 説明文を分割
        description_parts = split_text(video.description)
        blocks = []
        
        # 説明文ブロックの追加
        for part in description_parts:
            blocks.append({
                "object": "block",
                "type": "paragraph",
                "paragraph": {
                    "rich_text": [{"type": "text", "text": {"content": part}}]
                }
            })

        # 字幕見出しの追加
        blocks.append({
            "object": "block",
            "type": "heading_2",
            "heading_2": {
                "rich_text": [{"type": "text", "text": {"content": "字幕"}}]
            }
        })

        # 字幕ブロックの追加
        for caption in video.captions:
            # 言語情報の追加
            blocks.append({
                "object": "block",
                "type": "paragraph",
                "paragraph": {
                    "rich_text": [{
                        "type": "text",
                        "text": {"content": f"言語: {caption.language}"}
                    }]
                }
            })
            
            # 字幕テキストを分割して追加
            caption_parts = split_text(caption.text)
            for part in caption_parts:
                blocks.append({
                    "object": "block",
                    "type": "paragraph",
                    "paragraph": {
                        "rich_text": [{"type": "text", "text": {"content": part}}]
                    }
                })

        notion_client.pages.create(
            parent={"database_id": database_id},
            properties={
                "Title": {
                    "title": [{"text": {"content": video.title}}]
                },
                "URL": {
                    "url": video.url
                },
                "Channel": {
                    "multi_select": [{"name": video.channel_title}]
                }
            },
            children=blocks
        )

    except Exception as e:
        logging.error(f"Notion保存エラー: {e}")
        raise

def process_channel(youtube_service, notion_client, channel_id: str, database_id: str):
    """チャンネルの処理"""
    try:
        videos = get_latest_videos(youtube_service, channel_id)
        for video in videos:
            # get_captionsの呼び出しを修正
            video.captions = get_captions(video.video_id)  # youtube_serviceを渡さない
            save_to_notion(notion_client, database_id, video)
    except Exception as e:
        logging.error(f"チャンネル処理エラー: {e}")
def main():
    # ログ設定
    logging.basicConfig(
        level=logging.INFO,
        format='%(asctime)s [%(levelname)s] %(message)s'
    )

    # 環境変数の取得
    notion_api_key = os.getenv("NOTION_API_KEY")
    notion_database_id = os.getenv("NOTION_DATABASE_ID")

    if not all([notion_api_key, notion_database_id]):
        raise ValueError("Required environment variables are not set")

    # クライアントの初期化
    youtube_service = get_service_account_client()
    notion_client = Client(auth=notion_api_key)

    # チャンネルIDのリスト
    channel_ids = [
        "UCagAVZFPcLh9UMDidIUfXKQ",  # MBチャンネル
        "UC67Wr_9pA4I0glIxDt_Cpyw",  # 学長
        "UCXjTiSGclQLVVU83GVrRM4w",  # ホリエモン
    ]

    # チャンネルごとの処理
    for channel_id in channel_ids:
        process_channel(youtube_service, notion_client, channel_id, notion_database_id)

if __name__ == "__main__":
    main()

File: summary.go
package main

import (
	"io"
    "context"
    "fmt"
    "log"
    "os"
    "time"
    "strings"
    "sync"
    "github.com/jomei/notionapi"
	"google.golang.org/api/option"
    "google.golang.org/api/youtube/v3"
    "net/http"
	"encoding/json"
    "golang.org/x/oauth2"
    "golang.org/x/oauth2/google"
)

const (
    MaxDescriptionLength = 2000
    YouTubeScope = youtube.YoutubeReadonlyScope + " " + 
                  youtube.YoutubeForceSslScope
)

type VideoInfo struct {
    VideoID      string
    Title        string
    Description  string
    PublishedAt  time.Time
    ChannelTitle string
    URL          string
    Captions     []CaptionInfo
}

type CaptionInfo struct {
    Language    string
    Text        string
    IsAutomatic bool
}

// 説明文を制限する関数
func truncateDescription(description string) string {
    runes := []rune(description)
    if len(runes) > MaxDescriptionLength {
        return string(runes[:MaxDescriptionLength-3]) + "..."
    }
    return description
}

func main() {
    // 環境変数の取得
    notionAPIKey := os.Getenv("NOTION_API_KEY")
    notionDatabaseID := os.Getenv("NOTION_DATABASE_ID")
    if notionAPIKey == "" || notionDatabaseID == "" {
        log.Fatal("Required environment variables are not set")
    }

    ctx := context.Background()

    // サービスアカウント認証クライアントの取得
    client, err := getServiceAccountClient()
    if err != nil {
        log.Fatalf("サービスアカウントクライアントの作成に失敗: %v", err)
    }

    // YouTubeサービスの初期化
    youtubeService, err := youtube.NewService(ctx, option.WithHTTPClient(client))
    if err != nil {
        log.Fatalf("YouTubeサービスの作成に失敗: %v", err)
    }

    // Notionクライアントの初期化
    notionClient := notionapi.NewClient(notionapi.Token(notionAPIKey))

    // チャンネルIDのリスト
    channelIDs := []string{
        "UCagAVZFPcLh9UMDidIUfXKQ", // MBチャンネル
        "UC67Wr_9pA4I0glIxDt_Cpyw", // 学長
        "UCXjTiSGclQLVVU83GVrRM4w", // ホリエモン
    }

    // チャンネルごとの処理
    for _, channelID := range channelIDs {
        processChannel(ctx, youtubeService, notionClient, channelID, notionDatabaseID)
    }

}

// getServiceAccountClient関数の修正
func getServiceAccountClient() (*http.Client, error) {
    data, err := os.ReadFile("service-account.json")
    if err != nil {
        return nil, fmt.Errorf("サービスアカウントキーファイルの読み込みエラー: %v", err)
    }

    // 利用可能なスコープのみを指定
    creds, err := google.CredentialsFromJSON(context.Background(), data, 
        youtube.YoutubeReadonlyScope,
        youtube.YoutubeForceSslScope)
    if err != nil {
        return nil, fmt.Errorf("認証情報の作成エラー: %v", err)
    }

    return oauth2.NewClient(context.Background(), creds.TokenSource), nil
}

func tokenFromFile(file string) (*oauth2.Token, error) {
    f, err := os.Open(file)
    if err != nil {
        return nil, err
    }
    defer f.Close()
    tok := &oauth2.Token{}
    err = json.NewDecoder(f).Decode(tok)
    return tok, err
}

func getTokenFromWeb(config *oauth2.Config) *oauth2.Token {
    authURL := config.AuthCodeURL("state-token", oauth2.AccessTypeOffline)
    fmt.Printf("認証URLを開きます: \n%v\n", authURL)
    
    // 入力待ちであることを明確に表示
    fmt.Print("認証コードを入力してください: ")
    
    var authCode string
    if _, err := fmt.Scan(&authCode); err != nil {
        log.Printf("認証コードの読み取りに失敗: %v", err)
        return nil
    }
    
    // 入力された認証コードを表示
    log.Printf("入力された認証コード: %s", authCode)
    
    // コンテキストにタイムアウトを設定
    ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
    defer cancel()
    
    tok, err := config.Exchange(ctx, authCode)
    if err != nil {
        log.Printf("トークン取得に失敗: %v", err)
        return nil
    }
    
    log.Printf("トークン取得成功")
    return tok
}

func saveToken(path string, token *oauth2.Token) {
    fmt.Printf("Saving credential file to: %s\n", path)
    f, err := os.OpenFile(path, os.O_RDWR|os.O_CREATE|os.O_TRUNC, 0600)
    if err != nil {
        log.Fatalf("Unable to cache oauth token: %v", err)
    }
    defer f.Close()
    json.NewEncoder(f).Encode(token)
}

func getClient(config *oauth2.Config) *http.Client {
    tokFile := "token.json"
    tok, err := tokenFromFile(tokFile)
    if err != nil {
        // トークンファイルが存在しないか無効な場合のデバッグ出力
        log.Printf("トークンファイルの読み込みに失敗: %v", err)
        tok = getTokenFromWeb(config)
        saveToken(tokFile, tok)
    }
    return config.Client(context.Background(), tok)
}
// processChannel 関数の修正
func processChannel(ctx context.Context, youtubeService *youtube.Service, notionClient *notionapi.Client, channelID, databaseID string) {
    
    videos, err := getLatestVideos(youtubeService, channelID)
    if err != nil {
        log.Printf("エラー: チャンネル %s の動画取得に失敗: %v", channelID, err)
        return
    }
    log.Printf("チャンネル %s から %d 件の動画を取得しました", channelID, len(videos))

    var wg sync.WaitGroup
    semaphore := make(chan struct{}, 3)

    for _, video := range videos {
        wg.Add(1)
        go func(v VideoInfo) {
            defer wg.Done()
            semaphore <- struct{}{}
            defer func() { <-semaphore }()


            exists, err := checkDuplicateInNotion(notionClient, databaseID, v.VideoID)
            if err != nil {
                log.Printf("エラー: 重複チェック中 (VideoID: %s): %v", v.VideoID, err)
                return
            }
            if exists {
                log.Printf("スキップ: 動画 %s は既にNotionに存在します", v.VideoID)
                return
            }

            captions, err := getCaptions(youtubeService, v.VideoID)
            if err != nil {
                log.Printf("警告: 動画 %s の字幕取得に失敗: %v", v.VideoID, err)
            } else {
                log.Printf("字幕取得完了: %s (%d 件の字幕)", v.VideoID, len(captions))
            }
            v.Captions = captions

            err = saveToNotionWithRetry(notionClient, databaseID, v, 3)
            if err != nil {
                log.Printf("エラー: Notionへの保存失敗 (VideoID: %s): %v", v.VideoID, err)
                return
            }
        }(video)
    }

    wg.Wait()
}

// getLatestVideos 関数の修正
func getLatestVideos(service *youtube.Service, channelID string) ([]VideoInfo, error) {
    
    channelResponse, err := service.Channels.List([]string{"contentDetails"}).
        Id(channelID).
        Do()
    if err != nil {
        log.Printf("チャンネル情報取得エラー: %v", err)
        return nil, err
    }
    log.Printf("チャンネル情報取得成功")

    if len(channelResponse.Items) == 0 {
        return nil, fmt.Errorf("チャンネルが見つかりません")
    }

    uploadsPlaylistID := channelResponse.Items[0].ContentDetails.RelatedPlaylists.Uploads
    log.Printf("アップロードプレイリストID: %s", uploadsPlaylistID)

    var videos []VideoInfo
    playlistResponse, err := service.PlaylistItems.List([]string{"snippet"}).
        PlaylistId(uploadsPlaylistID).
        MaxResults(50).
        Do()
    if err != nil {
        return nil, fmt.Errorf("プレイリストアイテムの取得に失敗: %v", err)
    }

    now := time.Now().In(time.Local)
    today := time.Date(now.Year(), now.Month(), now.Day(), 0, 0, 0, 0, time.Local)
    yesterday := today.AddDate(0, 0, -1)

    filteredCount := 0
    for _, item := range playlistResponse.Items {
        publishedAt, err := time.Parse(time.RFC3339, item.Snippet.PublishedAt)
        if err != nil {
            log.Printf("警告: 動画 %s の日付解析に失敗: %v", item.Snippet.ResourceId.VideoId, err)
            continue
        }
        
        publishedAtJST := publishedAt.In(time.Local)
        publishedDate := time.Date(publishedAtJST.Year(), publishedAtJST.Month(), publishedAtJST.Day(), 0, 0, 0, 0, time.Local)

        if publishedDate.Equal(today) || publishedDate.Equal(yesterday) {
            video := VideoInfo{
                VideoID:      item.Snippet.ResourceId.VideoId,
                Title:        item.Snippet.Title,
                Description:  truncateDescription(item.Snippet.Description),
                PublishedAt:  publishedAt,
                ChannelTitle: item.Snippet.ChannelTitle,
                URL:         fmt.Sprintf("https://www.youtube.com/watch?v=%s", item.Snippet.ResourceId.VideoId),
            }
            videos = append(videos, video)
            filteredCount++
            log.Printf("対象動画を追加: %s (%s)", video.Title, video.PublishedAt.Format("2006-01-02 15:04:05"))
        }
    }

    log.Printf("チャンネル %s から %d 件の対象動画を抽出しました", channelID, filteredCount)
    return videos, nil
}
func getCaptions(service *youtube.Service, videoID string) ([]CaptionInfo, error) {
    captionResponse, err := service.Captions.List([]string{"snippet"}, videoID).Do()
    if err != nil {
        if strings.Contains(err.Error(), "forbidden") || 
           strings.Contains(err.Error(), "quotaExceeded") {
            log.Printf("警告: 動画 %s の字幕取得をスキップ: %v", videoID, err)
            return []CaptionInfo{}, nil
        }
        return nil, fmt.Errorf("字幕情報の取得エラー: %v", err)
    }

    var captions []CaptionInfo
    for _, caption := range captionResponse.Items {
        // 字幕テキストを取得
		resp, err := service.Captions.Download(caption.Id).Download()
		if err != nil {
			log.Printf("Error downloading caption: %v", err)
			continue
		}
		captionTrack, err := io.ReadAll(resp.Body)
		defer resp.Body.Close()
        captionInfo := CaptionInfo{
            Language:    caption.Snippet.Language,
            Text:        string(captionTrack),
            IsAutomatic: strings.Contains(caption.Snippet.TrackKind, "ASR"),
        }
        captions = append(captions, captionInfo)
    }

    return captions, nil
}

func checkDuplicateInNotion(client *notionapi.Client, databaseID, videoID string) (bool, error) {
    query := &notionapi.DatabaseQueryRequest{
        Filter: &notionapi.PropertyFilter{
            Property: "URL",
            RichText: &notionapi.TextFilterCondition{
                Contains: videoID,
            },
        },
    }
    
    result, err := client.Database.Query(context.Background(), notionapi.DatabaseID(databaseID), query)
    if err != nil {
        return false, err
    }
    
    return len(result.Results) > 0, nil
}

func saveToNotionWithRetry(client *notionapi.Client, databaseID string, video VideoInfo, maxRetries int) error {
    var lastErr error
    for i := 0; i < maxRetries; i++ {
        err := saveToNotion(client, databaseID, video)
        if err == nil {
            return nil
        }
        lastErr = err
        time.Sleep(time.Second * time.Duration(i+1))
    }
    return fmt.Errorf("failed after %d retries: %v", maxRetries, lastErr)
}


func saveToNotion(client *notionapi.Client, databaseID string, video VideoInfo) error {
    description := truncateDescription(video.Description)

	// ブロックの作成
	blocks := []notionapi.Block{
		&notionapi.ParagraphBlock{
			BasicBlock: notionapi.BasicBlock{
				Object: "block",
				Type:   notionapi.BlockTypeParagraph,
			},
			Paragraph: notionapi.Paragraph{
				RichText: []notionapi.RichText{
					{
						Type: "text",
						Text: &notionapi.Text{
							Content: description,
						},
					},
				},
			},
		},
		&notionapi.Heading2Block{
			BasicBlock: notionapi.BasicBlock{
				Object: "block",
				Type:   notionapi.BlockTypeHeading2,
			},
			Heading2: notionapi.Heading{
				RichText: []notionapi.RichText{
					{
						Type: "text",
						Text: &notionapi.Text{
							Content: "字幕",
						},
					},
				},
			},
		},
	}

	// 字幕ブロックの追加
	for _, caption := range video.Captions {
		blocks = append(blocks, &notionapi.ParagraphBlock{
			BasicBlock: notionapi.BasicBlock{
				Object: "block",
				Type:   notionapi.BlockTypeParagraph,
			},
			Paragraph: notionapi.Paragraph{
				RichText: []notionapi.RichText{
					{
						Type: "text",
						Text: &notionapi.Text{
							Content: fmt.Sprintf("言語: %s\n%s", caption.Language, caption.Text),
						},
					},
				},
			},
		})
	}
	
	
    params := &notionapi.PageCreateRequest{
        Parent: notionapi.Parent{
            Type:       notionapi.ParentTypeDatabaseID,
            DatabaseID: notionapi.DatabaseID(databaseID),
        },
        Properties: notionapi.Properties{
            "Title": notionapi.TitleProperty{
                Title: []notionapi.RichText{
                    {
                        Text: &notionapi.Text{
                            Content: video.Title,
                        },
                    },
                },
            },
            "URL": notionapi.URLProperty{
                URL: video.URL,
            },
            "Channel": notionapi.MultiSelectProperty{
                MultiSelect: []notionapi.Option{
                    {
                        Name: video.ChannelTitle,
                    },
                },
            },
        },
        Children: blocks,
    }

    _, err := client.Page.Create(context.Background(), params)
    return err
}
